File: backend/app/agent/normalizer.py
from __future__ import annotations

from app.models import Modality, NormalizedEvent

AUDIO_NOISE = (
    "cannot perform this analysis",
    "based on provided text",
    "i do not have access",
    "no access to the audio",
)


def _compact(text: str, max_len: int) -> str:
    clean = " ".join((text or "").split()).strip()
    if len(clean) <= max_len:
        return clean
    return clean[: max_len - 3] + "..."


def normalize_event(raw_event: dict) -> NormalizedEvent | None:
    channel = str(raw_event.get("channel") or "")
    data = raw_event.get("data") or {}

    if channel == "transcript":
        if not data.get("is_final"):
            return None
        text = _compact(str(data.get("text") or ""), 260)
        if not text:
            return None
        return NormalizedEvent(
            timestamp_start=float(data.get("start", 0.0)),
            timestamp_end=float(data.get("end", data.get("start", 0.0))),
            modality=Modality.TRANSCRIPT,
            text=text,
            is_final=True,
            raw_channel=channel,
        )

    if channel in {"scene_index", "visual_index"}:
        text = _compact(str(data.get("text") or ""), 380)
        if not text:
            return None
        return NormalizedEvent(
            timestamp_start=float(data.get("start", 0.0)),
            timestamp_end=float(data.get("end", data.get("start", 0.0))),
            modality=Modality.VISUAL,
            text=text,
            raw_channel=channel,
        )

    if channel == "audio_index":
        text = _compact(str(data.get("text") or ""), 280)
        raw_text = _compact(str(data.get("raw_text") or ""), 260)
        if not text:
            if raw_text:
                # Test-14 fallback: use audio raw transcript when transcript channel is sparse.
                return NormalizedEvent(
                    timestamp_start=float(data.get("start", 0.0)),
                    timestamp_end=float(data.get("end", data.get("start", 0.0))),
                    modality=Modality.TRANSCRIPT,
                    text=raw_text,
                    raw_channel="audio_index_raw_text",
                )
            return None
        low = text.lower()
        if any(p in low for p in AUDIO_NOISE):
            if raw_text:
                return NormalizedEvent(
                    timestamp_start=float(data.get("start", 0.0)),
                    timestamp_end=float(data.get("end", data.get("start", 0.0))),
                    modality=Modality.TRANSCRIPT,
                    text=raw_text,
                    raw_channel="audio_index_raw_text",
                )
            return None
        return NormalizedEvent(
            timestamp_start=float(data.get("start", 0.0)),
            timestamp_end=float(data.get("end", data.get("start", 0.0))),
            modality=Modality.AUDIO,
            text=text,
            raw_channel=channel,
        )

    if channel == "alert":
        label = _compact(str(data.get("label") or ""), 100)
        if not label:
            return None
        return NormalizedEvent(
            timestamp_start=float(data.get("start", 0.0)),
            timestamp_end=float(data.get("end", data.get("start", 0.0))),
            modality=Modality.ALERT,
            text=label,
            confidence=float(data.get("confidence", 1.0)),
            raw_channel=channel,
        )

    return None

File: backend-go/internal/agent/normalizer.go
package agent

import (
	"strings"

	"github.com/sangeetmore/novo-transcriber/backend-go/internal/models"
)

// audioNoisePhrases lists substrings that indicate an audio_index LLM response
// is noise rather than meaningful content — identical to the Python AUDIO_NOISE tuple.
var audioNoisePhrases = []string{
	"cannot perform this analysis",
	"based on provided text",
	"i do not have access",
	"no access to the audio",
}

// compact collapses all internal whitespace in text to single spaces, strips
// leading/trailing space, then truncates to maxLen characters, appending "..."
// if truncation was necessary. Mirrors Python _compact.
func compact(text string, maxLen int) string {
	clean := strings.Join(strings.Fields(text), " ")
	if len(clean) <= maxLen {
		return clean
	}
	return clean[:maxLen-3] + "..."
}

// NormalizeEvent converts a raw pipeline event map into a typed NormalizedEvent.
// It returns nil for events that should be discarded (non-final transcripts,
// audio noise without a raw fallback, empty text, or unrecognised channels).
// This is a direct port of the Python normalize_event function.
func NormalizeEvent(raw map[string]any) *models.NormalizedEvent {
	channel := ""
	if v, ok := raw["channel"]; ok && v != nil {
		channel, _ = v.(string)
	}

	var data map[string]any
	if v, ok := raw["data"]; ok && v != nil {
		data, _ = v.(map[string]any)
	}
	if data == nil {
		data = map[string]any{}
	}

	tsStart := toFloat(data["start"])
	tsEnd := toFloat(data["end"])
	if tsEnd == 0 {
		tsEnd = tsStart
	}

	switch channel {
	case "transcript":
		if !toBool(data["is_final"]) {
			return nil
		}
		text := compact(toString(data["text"]), 260)
		if text == "" {
			return nil
		}
		return &models.NormalizedEvent{
			TimestampStart: tsStart,
			TimestampEnd:   tsEnd,
			Modality:       models.ModalityTranscript,
			Text:           text,
			IsFinal:        true,
			RawChannel:     channel,
		}

	case "scene_index", "visual_index":
		text := compact(toString(data["text"]), 380)
		if text == "" {
			return nil
		}
		return &models.NormalizedEvent{
			TimestampStart: tsStart,
			TimestampEnd:   tsEnd,
			Modality:       models.ModalityVisual,
			Text:           text,
			IsFinal:        true,
			RawChannel:     channel,
		}

	case "audio_index":
		text := compact(toString(data["text"]), 280)
		rawText := compact(toString(data["raw_text"]), 260)

		if text == "" {
			if rawText != "" {
				// Test-14 fallback: use audio raw transcript when transcript channel is sparse.
				return &models.NormalizedEvent{
					TimestampStart: tsStart,
					TimestampEnd:   tsEnd,
					Modality:       models.ModalityTranscript,
					Text:           rawText,
					IsFinal:        true,
					RawChannel:     "audio_index_raw_text",
				}
			}
			return nil
		}

		low := strings.ToLower(text)
		for _, phrase := range audioNoisePhrases {
			if strings.Contains(low, phrase) {
				if rawText != "" {
					return &models.NormalizedEvent{
						TimestampStart: tsStart,
						TimestampEnd:   tsEnd,
						Modality:       models.ModalityTranscript,
						Text:           rawText,
						IsFinal:        true,
						RawChannel:     "audio_index_raw_text",
					}
				}
				return nil
			}
		}

		return &models.NormalizedEvent{
			TimestampStart: tsStart,
			TimestampEnd:   tsEnd,
			Modality:       models.ModalityAudio,
			Text:           text,
			IsFinal:        true,
			RawChannel:     channel,
		}

	case "alert":
		label := compact(toString(data["label"]), 100)
		if label == "" {
			return nil
		}
		confidence := 1.0
		if v, ok := data["confidence"]; ok {
			confidence = toFloat(v)
		}
		return &models.NormalizedEvent{
			TimestampStart: tsStart,
			TimestampEnd:   tsEnd,
			Modality:       models.ModalityAlert,
			Text:           label,
			Confidence:     confidence,
			IsFinal:        true,
			RawChannel:     channel,
		}
	}

	return nil
}

// ── type-coercion helpers ─────────────────────────────────────────────────────

func toString(v any) string {
	if v == nil {
		return ""
	}
	s, _ := v.(string)
	return s
}

func toFloat(v any) float64 {
	if v == nil {
		return 0
	}
	switch n := v.(type) {
	case float64:
		return n
	case float32:
		return float64(n)
	case int:
		return float64(n)
	case int64:
		return float64(n)
	}
	return 0
}

func toBool(v any) bool {
	if v == nil {
		return false
	}
	b, _ := v.(bool)
	return b
}
